Code/GraphMol/RGroupDecomposition/RGroupDecompParams.cpp

```diff
@@ -310,50 +315,66 @@ void RGroupDecompositionParameters::addDummyAtomsToUnlabelledCoreAtoms(
     return;
   }
 
-  // add a single rgroup substitution to any atom with free valence that doesn't
-  // have a terminal dummy neighbor
-
-  std::vector<Atom *> unlabelledCoreAtoms{};
+  // add R group substitutions to fill atomic valence
+  std::vector<Atom *> unlabeledCoreAtoms{};
   for (const auto atom : core.atoms()) {
     if (atom->getAtomicNum() == 1) {
       continue;
     }
-
     if (hasLabel(atom, labels)) {
       continue;
     }
+    unlabeledCoreAtoms.push_back(atom);
+  }
 
-    if (atom->getAtomicNum() > 0 && atom->calcImplicitValence() <= 1 &&
-        !atom->hasQuery()) {
-      continue;
-    }
+  for (const auto atom : unlabeledCoreAtoms) {
+    atom->calcImplicitValence(false);
+    const auto atomIndex = atom->getIdx();
+    int dummiesToAdd;
+    int maxNumDummies = 4 - static_cast<int>(atom->getDegree());
 
-    auto [nbrIdx, endNbrs] = core.getAtomNeighbors(atom);
-    bool hasTerminalDummyNeighbor = false;
+    // figure out the number of dummies to add
+    if (atom->getAtomicNum() == 0) {
+      dummiesToAdd = maxNumDummies;
+    } else {
+      double bondOrder = 0;
+      for (const auto bond : core.atomBonds(atom)) {
+        auto contrib = bond->getValenceContrib(atom);
+        if (contrib == 0.0 && bond->hasQuery()) {
+          contrib = 1.0;
+        }
+        bondOrder += contrib;
+      }
 
-    while (nbrIdx != endNbrs) {
-      const auto neighbor = core.getAtomWithIdx(*nbrIdx);
-      if (neighbor->getAtomicNum() == 0 && neighbor->getDegree() == 1) {
-        hasTerminalDummyNeighbor = true;
-        break;
+      auto valances =
+          PeriodicTable::getTable()->getValenceList(atom->getAtomicNum());
+      auto valence = *std::max_element(valances.begin(), valances.end());
```

Comment:
It feels like this is going to give weird/wrong answers for things like S and P, which have multiple possible valences, can be aromatic, and often appear in aromatic rings in a valence which is not their highest. For example `c1ccsc1 or `c1cccp1C`

Reply:
It shouldn't do.  The dummy atom does not have to match an RGroup, but any RGroup has to match a dummy atom.  If the core query contains `[#6][#16][#6]` then I add 2 dummies and it can match both `csc` (no R groups) and `CS(=O)(=O)C` (two `=O` R groups).
I try to add the smallest number of dummies here as there is an overhead to having unmatched dummies.

Reply:
Do we still have the same issue with S versus SH?  One will have an rgroup and one won’t.

Reply:
> Do we still have the same issue with S versus SH? One will have an rgroup and one won’t.

That will also work fine.  Extending the previous example, a core query of `[#6][#16][#6]` can also match `c[sH][c]` (assuming for the sake of argument such a thing could exist!).  In that case one dummy is mapped to an RGroup and the other is ignored.